tinygrad/renderer/cstyle.py

```diff
@@ -382,11 +377,15 @@ class AMDRenderer(CStyleLanguage):
             '__builtin_amdgcn_fence(__ATOMIC_ACQUIRE, "workgroup");'
   float4 = "make_float4"
   uses_ptr_arithmetic = False  # NOTE: this fixes TestLinearizerOverflowAlt
-  type_map = {dtypes.bfloat16: "hip_bfloat16"}
+  type_map = {dtypes.bfloat16: "hip_bfloat16", dtypes.half: "_Float16"}
+
+  def render_vector_prefix(self, dtype:DType) -> str:
+    vec, scal = self.render_dtype(dtype), self.render_dtype(dtype.scalar()),
+    return f"typedef {scal} {vec} __attribute__((ext_vector_type({dtype.count})));\nstatic inline __attribute__((device)) "+\
+           f"{vec} make_{vec}({', '.join([f'{scal} {x}' for x in _nms[:dtype.count]])}) {{ return {{ {', '.join(_nms[:dtype.count])} }}; }}"
 
   def render_kernel(self, function_name, kernel, bufs, uops, prefix=None) -> str:
     prefix = ["#define INFINITY (__builtin_inff())", "#define NAN (__builtin_nanf(\"\"))", "typedef long unsigned int size_t;"]
-    vec_dts = [("float", "float", 2), ("float", "float", 4), ("float", "float", 8), ("signed int", "int", 4), ("signed int", "int", 2)]
```

Comment:
One concern I have is on the fixed vector dtypes of `signed int` and `int`.
I believe this does not render the base_type as `signed int` and rather use only `int` for the vector dtype

Reply:
Unfortunately none of the process_replay kernels use vectorized int so I can't tell if it works. Don't know if CI test it either

Reply:
you can check CI coverage by removing stuff and see if any test breaks. if not, can you add something that triggers these (likely in test_linearizer, construct a kernel and apply different upcast to it)

i pushed this to internal benchmark that runs on real AMD

Reply:
oh we only vectorize half and float now I think, there's no int4 after lowerer.

Reply:
`test_dtype/test_upcast_ops` generates the following for `METAL` backend. Guess no vectorization for `int` dtypes, it just seems to unroll.

``` c
#include <metal_stdlib>
using namespace metal;
kernel void E_4n41(device int* data0, const device short* data1, const device int* data2, uint3 gid [[threadgroup_position_in_grid]], uint3 lid [[thread_position_in_threadgroup]]) {
  short val0 = *(data1+0);
  short val1 = *(data1+1);
  short val2 = *(data1+2);
  short val3 = *(data1+3);
  int val4 = *(data2+0);
  int val5 = *(data2+1);
  int val6 = *(data2+2);
  int val7 = *(data2+3);
  *(data0+0) = ((int)(val0)*val4);
  *(data0+1) = ((int)(val1)*val5);
  *(data0+2) = ((int)(val2)*val6);
  *(data0+3) = ((int)(val3)*val7);
}
# --- #
#include <metal_stdlib>
using namespace metal;
kernel void E_4n104(device unsigned int* data0, const device unsigned short* data1, const device unsigned int* data2, uint3 gid [[threadgroup_position_in_grid]], uint3 lid [[thread_position_in_threadgroup]]) {
  unsigned short val0 = *(data1+0);
  unsigned short val1 = *(data1+1);
  unsigned short val2 = *(data1+2);
  unsigned short val3 = *(data1+3);
  unsigned int val4 = *(data2+0);
  unsigned int val5 = *(data2+1);
  unsigned int val6 = *(data2+2);
  unsigned int val7 = *(data2+3);
  *(data0+0) = ((unsigned int)(val0)+val4);
  *(data0+1) = ((unsigned int)(val1)+val5);
  *(data0+2) = ((unsigned int)(val2)+val6);
  *(data0+3) = ((unsigned int)(val3)+val7);
}
```

Reply:
> i pushed this to internal benchmark that runs on real AMD

did it work?